packages/cube/pipeline/cube-validation.ttl

```diff
@@ -0,0 +1,229 @@
+  p:value 3 ;
+.
+
+
+<fetch-cube-constraint> a p:Pipeline ;
+  p:variables [ p:variable _:endpoint, _:cube ] ;
+  p:steps
+    [
+      p:stepList
+        (
+          _:queryConstraint
+          [ ntriples:serialize () ]
+          _:stdout
+        ) 
+    ] 
+.
+
+<fetch-cube-observations> a p:Pipeline ;
+  p:variables [ p:variable _:endpoint, _:cube ] ;
+  p:steps
+    [
+      p:stepList
+        (
+          _:queryObservations
+          [ ntriples:serialize () ]
+          _:stdout
+        ) 
+    ] 
+.
+
+<check-cube-constraint> a p:Pipeline ;
+  p:variables [ p:variable _:profile ] ;
+  p:steps
+    [
+      p:stepList ( 
+        _:stdin
+        [ n3:parse () ]
+        _:toDataset
+        [ shacl:report (_:getProfile) ]
+        [ base:flatten () ]
+        [ ntriples:serialize () ]
+        _:stdout
+      )
+    ]
+.
+
+<check-cube-observations> a p:Pipeline ;
+  p:variables [ p:variable _:constraint , _:batchSize , _:maxViolations ] ;
+  p:steps
+    [
+      p:stepList ( 
+        _:stdin
+        [ n3:parse () ]
+        _:sortBySubject
+        _:toDatasetBySubject
+        _:batch
+        [ shacl:report 
+          [code:name "shape" ; code:value _:getConstraint ] ,
+          [code:name "maxErrors" ; code:value "maxViolations"^^p:VariableName ] 
+        ]
+        [ base:flatten () ]
+        [ ntriples:serialize () ]
+        _:stdout
+      )
+    ]
+.
+
+# relying on cbd is not ideal (vendor specific)
+_:queryConstraint sparql:construct
+    [ code:name "endpoint"; code:value "endpoint"^^p:VariableName ] , 
+    [
+      code:name "query";
+      code:value """
+      #pragma describe.strategy cbd
+
+      PREFIX cube: <https://cube.link/>
+
+      DESCRIBE ?s
+      WHERE {
+        <${cube}> cube:observationConstraint ?s .
+      }
+      """^^code:EcmaScriptTemplateLiteral 
+    ]
+.
+
+_:queryObservations sparql:construct
+    [
+      code:name "endpoint";
+      code:value "endpoint"^^p:VariableName
+    ] , 
+    [
+      code:name "query";
+      code:value """
+      PREFIX cube: <https://cube.link/>
+
+      CONSTRUCT { ?s ?p ?o }
+      WHERE {
+        <${cube}> cube:observationSet/cube:observation ?s .
+        ?s ?p ?o 
+      }
+      """^^code:EcmaScriptTemplateLiteral 
+    ]
+.
+
+_:stdin a p:Step ;
+  code:implementedBy "() => process.stdin"^^code:EcmaScript  
+.
+
+_:stdout a p:Step ;
+  code:implementedBy "() => process.stdout"^^code:EcmaScript  
+.
+
+_:toDataset a p:Step ;
+	code:implementedBy 
+    [
+      rdf:type code:EcmaScript ;
+      code:link <node:rdf-stream-to-dataset-stream/toDatasetStream.js> ;
+    ] 
+.
+
+_:toDatasetBySubject a p:Step ;
+	code:implementedBy 
+    [
+      rdf:type code:EcmaScript ;
+      code:link <node:rdf-stream-to-dataset-stream/bySubject.js> ;
+    ] 
+. 
+
+_:batch a p:Step ;
+  code:implementedBy 
+    [ 
+      a code:EcmaScriptModule ;
+      code:link <file:../lib/batch.js#batch> ;
+    ] ;
+  code:arguments ("batchSize"^^p:VariableName)
+.
```

Comment:
Isn't there some step like this in `barnard59-base`?

Reply:
apparently not, in fact I would consider adding one (or even two: one for datasets and another for generic objects)